apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx

```diff
@@ -0,0 +1,296 @@
+import {
+  keepPreviousData,
+  useInfiniteQuery,
+  useQuery,
+  useQueryClient,
+} from '@tanstack/react-query';
+import { useReducer } from 'react';
+
+import { commentErrors, sortKeys } from './comments.constants';
+import { getPaginatedComments } from './getCommentRouteData';
+import type { CommentRoot } from '@repo/db/types';
+import {
+  addComment as addCommentAction,
+  deleteComment as deleteCommentAction,
+  replyComment,
+  updateComment as updateCommentAction,
+} from './comment.action';
+import { toast } from '@repo/ui/components/use-toast';
+
+const getRootQueryKey = (rootId: number, type: CommentRoot) =>
+  `${type.toLowerCase()}-${rootId}-comments`;
+
+export type SortItem = (typeof sortKeys)[number];
+
+interface CommentsMeta {
+  page: number;
+  sort: SortItem;
+}
+
+interface DefaultCommentsProps {
+  rootId: number;
+  type: CommentRoot;
+}
+
+interface UseCommentsProps extends DefaultCommentsProps {
+  initialPage?: number;
+}
+
+export function useComments({ type, rootId, initialPage }: UseCommentsProps) {
+  const queryClient = useQueryClient();
+  const [meta, updateMeta] = useReducer(
```

Comment:
thought: this file reads great, but I did have to scratch my head a bit to figure out what `meta` is and where we create that. For a minute I thought I was looking at [meta](https://tanstack.com/query/v4/docs/react/guides/query-functions#queryfunctioncontext) from @tanstack/query.